MH-10560 Live Scheduler Service

rute-santos opened the pull request:
This service creates/updates a LIVE media package in the search index so that a student can watch a lecture while it's happening. 

The service listens to Scheduler / Asset Manager messages and updates the live media package accordingly.

The admin UI shows the Engage Live publication channel and allows launching the player while the capture is happening.

The capture agent used needs to be capable of streaming and the player needs to be capable of playing the format streamed.

Harvard DCE has been using the 1.x version of this code for years. We use Epiphan Pearls streaming rtmp and Paella (rtmp support via Flash). We haven't tested it with other capture agents, streaming formats, or players. Suggestions are welcome!

UI changes:

When event is scheduled and before start time, playing is disabled:

![events_view_scheduled](https://user-images.githubusercontent.com/9451622/34688930-c6c421d0-f481-11e7-8214-7c477b366d0b.png)
![event_details_scheduled](https://user-images.githubusercontent.com/9451622/34688944-d2377cec-f481-11e7-998f-8e80e84e8728.png)

While event is being captured, player can be launched:

![event_details_capturing](https://user-images.githubusercontent.com/9451622/34688997-02d58718-f482-11e7-84d7-b7ea5238e65c.png)

And in the Media Module:

![media_module_capturing](https://user-images.githubusercontent.com/9451622/34689012-10ef7296-f482-11e7-8084-a149a111042e.png)

rute-santos committed:
MH-10560-live-schedule Publish live mp to search index and add pub channel to asset. Listen to scheduler and asset manager events and updates live mp accordingly. UI shows live event, which can be played only after capture has started.

smarquard commented:
Hi @rute-santos, some conflicts have arisen from other merges to develop. Please could you resolve them.

I'll be reviewing this in the next few days

rute-santos commented:
Hi @smarquard, I resolved the conflicts and rebased to keep things clean. Thanks!

JanKoppe commented:
Hello Rute,

we're (WWU Münster) very interested in this feature. At a first glance I couldn't find any documentation on what the capture agent needs to implement to support this feature, i.e. some API documentation and workflow explanation. Can you point me to the information, or maybe add this to the PR?

Regards,
Jan

rute-santos commented:
Hi @JanKoppe,
This feature can only be used if the capture agent is already capable of streaming and the CA configuration will be different for each one. The idea is that the capture agent will stream to a server and the LiveSchedulerService will be configured to use the same url in the tracks. You also need to use  the Paella player. Which capture agent do you currently use?
Thanks,
Rute

JanKoppe commented:
Hello @rute-santos, yes, that's obvious, but I am specifically interested in which API the capture agent needs to implement, so I can add this feature to pyCA, or develop a stand alone program for this.

Just wanted to ask if there is any documentation about this at all, or people wanting to implement this have to read the entire code and figure it out for themselves (possibly missing intended behaviour).

Regards

rute-santos committed:
MH-10560-live-schedule Added additional configuration to documentation
for the case where the CA registers the streaming url.

rute-santos commented:
Hi @JanKoppe,
The capture agent doesn't _need_ to implement a specific api if the service is configured with the same url as the capture agent streams to.
But evidently this is not ideal because the configuration in both places need to match and all capture agents need to follow the same pattern so here at Harvard-DCE our capture agents pass the streaming urls when they register to Opencast and the LIveSchedulerService uses those urls when building the live tracks.
I've just added that to the documentation. Please refer to docs/guides/admin/docs/modules/liveschedule.md and make suggestions for improvements.
I also hope you don't need to read the code! :)
Thanks for your comments!
Rute

JanKoppe commented:
Ah, I think that clears it up a bit - I was thinking that the livestreaming is handled over a new API endpoint, but instead it extends the existing recording calendars endpoint.

Will take a look if we can extend pyCA to include this (would need to parse the attachments which we don't do yet), but one could just go ahead and push the stream every time a recording starts, but Opencast would only publish the event to engage when it actually has been marked with "publishLive", correct?

Lookin forward to playing with this, thank you! 👍

rute-santos committed:
MH-10560-live-schedule Fixed inconsistency in logger used.

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+
+The live media package is retracted from the Search index when the capture finishes or if it fails.
+
+If event metadata are updated, such as title, duration, the live media package in the Search index is updated accordingly.
```
Would be clearer to say:

```
If event metadata such as title or duration are updated, the live media package in the Search index is updated accordingly.
```

rute-santos replied:
Yes!

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+
+Edit  _etc/org.opencastproject.liveschedule.impl.LiveScheduleServiceImpl.cfg_.
+
+If you capture agent does not register a _capture.device.live.resolution.WIDTHxHEIGHT_ property, it's mandatory to configure the _live.streamingUrl_. 
```
you > your. Shouldn't this be mandatory to configure live.resolution? I don't get the connection between capture.device.live.resolution.WIDTHxHEIGHT and live.streamingUrl ?

rute-santos replied:
The live.resolution has a default value and that's why it isn't mandatory.
There's no connection between the capture.device.live.resolution.WIDTHXHEIGHT and the live.streamingUrl. You either use one or the other i.e. if the capture agent registers with the first, the other is ignored for that capture agent. 
It seems the documentation is confusing :)

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+# The streaming base url e.g. rtmp://streaming.server/live/
+#live.streamingUrl=rtmp://streaming.server/live
+
+# If a comma-separated list is informed, several resolutions will be generated for each flavor
```
I think "is provided" rather than "is informed"

rute-santos replied:
Ok!

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+# The same mime-type applies to all flavors and resolutions
+live.mimeType=video/x-flv
+
+# If a comma-separated list is informed, several streams links will be generated, one for each
```
informed change to provided

rute-santos replied:
Ok!

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+
+#### Capture agent registers the _capture.device.live.resolution.WIDTHxHEIGHT_ property 
+
+If you develop the code that runs on your capture agent, instead of configuring the live.streamingUrl,
```
I think it would be clearer to say:

"If your capture agent supports configuring custom capture agent properties, ..."

rute-santos replied:
Yes!

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+
+####Capture agent registers with capture.device.live.resolution.WIDTHxHEIGHT
+
+If the capture agent registers itself with:
```
I think it would be clearer to just put this in a table (property key, value) rather than the XML which adopters won't necessarily be familiar with or care about.

rute-santos replied:
Ok!

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+### Step 3: Configure the Workflow
+
+When scheduling a live event via the admin UI, the workflow needs to have the _publishLive_ configuration set to true (this is already included in the sample workflows).
+If not using the sample opencast workflows, add to the `<configuration_panel>`:
```
Opencast (capitalize)

rute-santos replied:
Good catch.

smarquard commented on etc/org.opencastproject.liveschedule.impl.LiveScheduleServiceImpl.cfg:
```diff
@@ -0,0 +1,34 @@
+# The streaming base url e.g. rtmp://streaming.server/live/
+#live.streamingUrl=rtmp://streaming.server/live
+
+# If a comma-separated list is informed, several resolutions will be generated for each flavor
```
"is provided"

rute-santos replied:
Done.

smarquard commented on etc/org.opencastproject.liveschedule.impl.LiveScheduleServiceImpl.cfg:
```diff
@@ -0,0 +1,34 @@
+# The same mime-type applies to all flavors and resolutions
+live.mimeType=video/x-flv
+
+# If a comma-separated list is informed, several streams links will be generated, one for each
```
"is provided"

rute-santos replied:
Done.

smarquard commented on modules/admin-ui/src/main/webapp/scripts/shared/resources/eventsResource.js:
```diff
@@ -42,6 +42,15 @@ angular.module('adminNg.resources')
                 row.publications = r.publications;
+                if (typeof(r.publications) != 'undefined' && r.publications != null) {
+                	var now = new Date();
+                	console.log("now: " + now + ", start: " + new Date(r.start_date) + ", end: " + new Date(r.end_date));
```
maybe remove this?

rute-santos replied:
Oops!

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/impl/LiveScheduleServiceImpl.java:
```diff
@@ -0,0 +1,818 @@
+ * Implement {@link SmilService} interface.
+ */
+public class LiveScheduleServiceImpl implements LiveScheduleService {
+  /** The eserver url property **/
```
eserver > server

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/impl/LiveScheduleServiceImpl.java:
```diff
@@ -0,0 +1,818 @@
+  private static final String DEFAULT_LIVE_TARGET_FLAVORS = "presenter/delivery";
+  static final String DEFAULT_LIVE_DISTRIBUTION_SERVICE = "download";
+
+  /** For MHPearl capture agent only */
```
Could be any CA that supports custom configuration, so rather leave this line out (that is, don't mention specific product)

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/impl/LiveScheduleServiceImpl.java:
```diff
@@ -0,0 +1,818 @@
+      MediaPackage newMp = replaceAndDistributeAcl(previousMp, acl);
+      // Publish mp to engage search index
+      publish(newMp);
+      // Don't live garbage there!
```
live > leave

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/impl/LiveScheduleServiceImpl.java:
```diff
@@ -0,0 +1,818 @@
+    // Publish mp to engage search index
+    publish(mp);
+    // Publication channel already there so no need to add
+    // Don't live garbage there!
```
live > leave

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/message/LiveScheduleMessageReceiver.java:
```diff
@@ -0,0 +1,175 @@
+import java.util.concurrent.locks.Lock;
+
+/**
+ * Very simple approach to serialize the work of all three dependend update handlers. Todo: Merge all handlers into one
```
dependent

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/message/LiveScheduleMessageReceiver.java:
```diff
@@ -0,0 +1,175 @@
+
+  private static final Logger logger = LoggerFactory.getLogger(LiveScheduleMessageReceiver.class);
+  // Striped lock for synchronizing on media package
+  private static final Striped<Lock> lock = Striped.lazyWeakLock(1);
```
This was changed where it was elsewhere in the code to Striped.lazyWeakLock(1024) to avoid unnecessary lock contention (when starting tasks), so probably a good idea to change it here too.

rute-santos replied:
Thanks!

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/message/SchedulerUpdateHandler.java:
```diff
@@ -0,0 +1,133 @@
+      Map<String, String> config = schedulerService.getWorkflowConfig(mpId);
+      return BooleanUtils.toBoolean((String) config.get(PUBLISH_LIVE_PROPERTY));
+    } catch (UnauthorizedException | NotFoundException | SchedulerException e) {
+      logger.info("Could not get workflow configuration for mp {}. This is probably ok.", mpId);
```
Maybe make this is a debug?

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+
+Note: Please refer to your streaming server or CDN documentation for the correct syntax of the streaming url.
+The _live.streamingUrl_ may be very different from the url the capture agent streams to.
+For instance, with Akamai, the url used by the player will be something like live.streamingUrl=rtmp://xyz.live.edgefcs.net/live/ and the capture agent's publish url something like rtmp://a.bcd.e.akamaientrypoint.net/EntryPoint. The stream name should always match.
```
Where is the capture agent's publish url configured? Only on the capture agent? Does Opencast never know about it?

rute-santos replied:
Only on the capture agent.

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,169 @@
+live.resolution, live.streamName, 
+you can update the capture agent firmware to pass the following when registering to Opencast:
+
+- capture.device.names: add 'live' to the current list of devices
```
The code doesn't appear to use this at all, as it relies on the publishLive workflow configuration property. Is there any requirement for a capture agent do include "live" as a device then?

rute-santos replied:
You are correct. The LiveScheduler code does not use it. We included it for consistency, because all the capture agent properties have the 'device name' in them e.g. capture.device.names=presenter and capture.device.presenter.flavor=presenter/source. Honestly, I don't know if this is necessary or not.

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/impl/LiveScheduleServiceImpl.java:
```diff
@@ -0,0 +1,818 @@
+/**
+ * Implement {@link SmilService} interface.
+ */
+public class LiveScheduleServiceImpl implements LiveScheduleService {
```
Not strictly necessary, but it would be nice to implement updated(...) so that changes to org.opencastproject.liveschedule.impl.LiveScheduleServiceImpl.cfg
 are updated live.

rute-santos replied:
You are correct. If it's ok with you I will add a todo comment and do it at a later date.

smarquard commented:
I'm not seeing anything show up under Publications in the event popup dialog under General before or during the live event.

The publications column is correct in the events table, and it shows the correct status before and during the live event.

wsmirnow commented:
Hi @smarquard, the activemq configuration was changed by this PR. Do you updated your activemq config and restarted the service? Sorry for the question, just to be sure.

smarquard commented:
@rute-santos, were you able to reproduce the issue in the Admin UI not showing the publication in the event popup modal?

rute-santos commented:
Hi @smarquard, I was on vacation last week and I will take a look this week.

rute-santos committed:
MH-10560-live-schedule Fixed bug caused by merge where event details would not show publications

rute-santos commented:
Hi @smarquard, I've fixed the bug. Thanks!

rute-santos committed:
MH-10560-live-schedule Rebased to current develop and fixed resulting errors (workflows method removed, style of live msg).

rute-santos committed:
MH-10560-live-schedule Removed snapshot ownership as a means to know who created the snapshot. Using a local cache instead.

rute-santos committed:
MH-10560-live-schedule Added changes from PR #129. Added changes requested by reviewer.

rute-santos commented:
Fixed the code that got broken because of PR #43. This was tested with PR #129 applied so it may not work without it.
And please let me know if you prefer that I squash all the commits into one!

rute-santos commented:
Hi @smarquard, Just checking: is this going to be merged into the new release or do I have to do anything else? Thanks!

smarquard commented:
@rute-santos, I will give it another test now and certainly hope to merge before feature freeze.

Could you look through the comments from a while back which I added, specifically some wording changes suggested for documentation. Also a couple of code comments.

rute-santos commented:
@smarquard, I cannot find the comments you mentioned.

smarquard commented:
Sorry, it seems github kept them all as pending until I submitted a complete review which I've done now. I didn't realize that made them invisible to anyone else.

rute-santos committed:
MH-10560-live-schedule Changes based on reviewer comments.

rute-santos commented:
Hi @smarquard,
I am glad I am not crazy :) 
I believe I addressed all your comments now. Please let me know if other changes are needed and thank you so much for reviewing!
Rute

smarquard commented:
After testing the latest build:
* The mediapackage is published as expected to the media module on scheduling
* In the media module, the event shows up as "Live (in progress)" before the scheduled start date. Not sure why, as the code looks ok to me. Maybe something wrong with the date calculations or comparison.
* When the event ends, the event is retracted from the media module as expected.
* Nothing is showing up in the Admin UI related to the live event, i.e. it's not shown with a live publication at all either in the events table or in the event popup under Publications.

smarquard commented:
![image](https://user-images.githubusercontent.com/5888197/38142164-a4978242-343b-11e8-8d1c-f305176d37b2.png)

All of these events are coming up, but some are shown as in progress and one not.

Also in the Admin UI, the publication status for a forthcoming live event shows up if you edit the metadata after the initial creation, but not otherwise.

smarquard commented:
![image](https://user-images.githubusercontent.com/5888197/38142270-f990364a-343b-11e8-922b-454ee325fcc6.png)

smarquard commented on docs/guides/admin/docs/modules/liveschedule.md:
```diff
@@ -0,0 +1,164 @@
+|-------------|-----|
+|capture.device.names|presentation,presenter,live|
+|capture.device.presentation.flavor|presentation/source|
+|capture.device.presentater.flavor|presenter/source|
```
Cosmetic here, but should probably be capture.device.presenter.flavor

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/impl/LiveScheduleServiceImpl.java:
```diff
@@ -0,0 +1,827 @@
+      liveStreamingUrl = StringUtils.trimToEmpty((String) properties.get(LIVE_STREAMING_URL));
+      logger.info("Live streaming server url is {}", liveStreamingUrl);
+    } else {
+      logger.warn("Live streaming url was not set in '{}'. This is a problem when not using MHPearl capture agents!",
```
Please remove reference reference to MHPearl here - perhaps something like

+      logger.info("Live streaming url not set in '{}'. Streaming urls must be provided by capture agent properties.")

(suggesting info rather than warn because it's not necessarily a config error).

smarquard commented on modules/live-schedule-impl/src/main/java/org/opencastproject/liveschedule/impl/LiveScheduleServiceImpl.java:
```diff
@@ -0,0 +1,827 @@
+  void addLiveTracks(MediaPackage mp, String caName) throws LiveScheduleException {
+    String mpId = mp.getIdentifier().compact();
+    try {
+      // MHPearl specific: if capture agent registered the properties:
```
Remove "MHPearl specific", i.e. just

      // If capture agent registered the properties:

rute-santos commented:
Hi @smarquard, I think the publication is not showing up because PR #129 is missing. I made a comment that it was needed, but it got lost among all the other comments. I will include it with this PR and make sure all works.

rute-santos commented:
Hi @smarquard, 
I addressed everything, except the media module issue. The issue is happening because, when the event is scheduled, the media package 'start' field is set to the creation date/time and not the capture start date/time. The live scheduler service gets the media package from the asset manager, which has the incorrect 'start' date/time. I don't know if I will be able to fix that before the release cut date, but I can provide a bug fix after that if you decide to merge.
Thanks!

smarquard commented:
Thanks @rute-santos 

github is reporting a conflict in AbstractAssetManager.java; can you resolve that?

rute-santos committed:
Merge remote-tracking branch 'community/develop' into f/MH-10560-live-schedule

rute-santos commented:
I've merged the current develop and resolved the conflict.

smarquard commented:
Merged - thank you for the contribution!

Please create a follow-up JIRA and PR for the issue in the media module where the mediapackage start date is incorrect: appears to be the creation date/time of the item, and should be the capture start date/time which I assume could differ from dcCreated as the 'descriptive' start date/time. The live stream should follow the schedule start/date time.

Please also look at the behaviour when rebuilding the Admin UI index; it looks like the live schedule service is trying to do something for past live scheduled events when a reindex event is received.
